File: Graph/Java/Minimum Score of a Path Between Two Cities.java
// Problem link: https://leetcode.com/problems/minimum-score-of-a-path-between-two-cities/description/
// Time: O(n)
// Space: O(n)
class Solution {
    public int minScore(int n, int[][] roads) {
        List<int[]>[] adjList = new ArrayList[n + 1];
        boolean[] visit = new boolean[n + 1];

        for(int idx = 0; idx < n + 1; idx++) {
            adjList[idx] = new ArrayList<>();
        }

        for(int[] road : roads) {
            adjList[road[0]].add(new int[] {road[1], road[2]});
            adjList[road[1]].add(new int[] {road[0], road[2]});
        }

        return this.dfs(1, adjList, visit, Integer.MAX_VALUE);
    }

    private int dfs(int node, List<int[]>[] adjList, boolean[] visit, int score) {
        visit[node] = true;

        for(int[] pair : adjList[node]) {
            if(!visit[pair[0]]) {
                score = Math.min(score, Math.min(pair[1], this.dfs(pair[0], adjList, visit, score)));
            }
            else {
                score = Math.min(score, pair[1]);
            }
        }

        return score;
    }
}

File: Graph/Python/Minimum Score of a Path Between Two Cities.py
# Problem link: https://leetcode.com/problems/minimum-score-of-a-path-between-two-cities/description/
# Time: O(n)
# Space: O(n)
class Solution:
    def minScore(self, n: int, roads: List[List[int]]) -> int:
        adjList = [[] for _ in range(n + 1)]
        visit = [False] * (n + 1)

        for u, v, w in roads:
            adjList[u].append([v, w])
            adjList[v].append([u, w])

        return self.dfs(1, visit, adjList, float('inf'))
    
    def dfs(self, node, visit, adjList, score):
        visit[node] = True

        for nei, wei in adjList[node]:
            if not visit[nei]:
                score = min(wei, score, self.dfs(nei, visit, adjList, score))
            else:
                score = min(score, wei)

        return score
